# Choose a Data Set

Create your own dataset by scraping one of the following websites *(level 5)*:
- [Wikipedia](https://www.wikipedia.org/)
- [OpenLibrary](https://openlibrary.org/)

**OR** 

Use data gathered from one of the following APIs *(level 4)*: 
- [TMDB](https://developer.themoviedb.org/reference/intro/getting-started)
- [College Scorecard](https://collegescorecard.ed.gov/data/api-documentation/)

**OR** 

Pick a JSON dataset *(level 3)*:
- [Food/Restaurant Data](https://drive.google.com/drive/folders/1V94S6WpclvQmbnW88KVMD4EruryA1oma?usp=drive_link)
- [Fashion Data](https://drive.google.com/drive/folders/1V8SbFjtRRW8WVf3xBzg0gzLjOtMhHea_?usp=drive_link)

**OR** 

Pick a CSV dataset *(level 2)*:
- [LA Parking Tickets](https://drive.google.com/drive/folders/1vaOfwMi6QmZEGsXr8VM0ulPGzvTTBCgm?usp=drive_link)
- [Hotels](https://drive.google.com/drive/folders/1IpVFxgwBJvJHKoOuBsk6WK2qYqFYP4hi?usp=drive_link)

# My Question
### With this list of Generation 1 Pokemon, what is the probability of choosing a pokemon that is a fire type? What about Grass and Poison? Finally what's the probability of you choosing an electric OR water type?

In [413]:
from bs4 import BeautifulSoup as BS  
import requests  
import pandas as pd
import random

In [319]:
url = 'https://en.wikipedia.org/wiki/List_of_generation_I_Pok%C3%A9mon'  
response = requests.get(url)  
html_stuff = response.text  
soup = BS(response.content, "html.parser")  

In [320]:
table = soup.find('table', {'class': 'wikitable'})   
rows = table.find_all('tr')  

In [323]:
headers = [th.text.strip() for th in rows[0].find_all('th')]  

In [325]:
data = []   
for row in rows[1:]:    
  cols = row.find_all(['th', 'td'])  # Find both th and td tags   
  row_data = []  
  for col in cols:  
     if col.text.strip() == '':  
        row_data.append('N/A')  # Handle empty entries  
     else:  
        row_data.append(col.text.strip())  
  data.append(row_data) 

In [327]:
max_cols = max(len(row) for row in data)  

In [329]:
while len(headers) < max_cols:  
   headers.append('Extra Column') 

In [331]:
for row in data:  
   while len(row) < max_cols:  
      row.append('N/A')  

In [333]:
df = pd.DataFrame(data, columns=headers)  

In [407]:
df = df.iloc[:, [4, 2, 3, 0, 1, 5]]  #WHY in God's time alive is this line of code messing EVERYTHING up from time to time?

In [409]:
df.columns = ['Name[nb 2]', 'Type(s)', 'Evolves from', 'Evolves into', 'Notes', 'Extra Column']  
df 

Name[nb 2]               Type(s)  \
0      BulbasaurFushigidane (フシギダネ)(0001)        Grass / Poison   
1          IvysaurFushigisō (フシギソウ)(0002)        Grass / Poison   
2       VenusaurFushigibana (フシギバナ)(0003)        Grass / Poison   
3         CharmanderHitokage (ヒトカゲ)(0004)                  Fire   
4           CharmeleonRizādo (リザード)(0005)                  Fire   
..                                    ...                   ...   
147        DragonairHakuryū (ハクリュー)(0148)                Dragon   
148         DragoniteKairyū (カイリュー)(0149)       Dragon / Flying   
149            MewtwoMyūtsū (ミュウツー)(0150)               Psychic   
150                    MewMyū (ミュウ)(0151)               Psychic   
151  MissingNo.Ketsuban (けつばん)(000[nb 9])  Normal / Bird[nb 10]   

           Evolves from                                       Evolves into  \
0                     —                                    Ivysaur (#0002)   
1     Bulbasaur (#0001)                                   Venusaur (#0003)   
2       Ivysaur (#0002)                           Mega EvolutionGigantamax   
3                     —                                 Charmeleon (#0005)   
4    Charmander (#0004)                                  Charizard (#0006)   
..                  ...                                                ...   
147     Dratini (#0147)                                  Dragonite (#0149)   
148   Dragonair (#0148)                                                  —   
149                   —                                Two Mega Evolutions   
150        No evolution  This mythical Pokémon is so rare, only a few e...   
151        No evolution  An error handler whose name stands for "Missin...   

                                                 Notes Extra Column  
0    It is one of Kanto's starter Pokémon. It has a...          N/A  
1    It is a playable character in Super Smash Bros...          N/A  
2    It is the mascot of Pokémon Green and LeafGree...          N/A  
3    It is one of Kanto's starter Pokémon. From the...          N/A  
4    A feisty Pokémon, it seeks out potential oppon...          N/A  
..                                                 ...          ...  
147  Dragonair can grow the wings on its head to fl...          N/A  
148  Capable of circling the globe in just sixteen ...          N/A  
149  It is a playable character in Super Smash Bros...          N/A  
150                                                N/A          N/A  
151                                                N/A          N/A  

[152 rows x 6 columns]

# Fire Type Empirical Probability

In [453]:
rounds = 0
FType = 0
trials = 10000
while rounds < 10000:
    num = random.randint(1,152)
    if num <= 12:
        FType += 1 
    rounds += 1

print(str((FType/trials) * 100))

7.670000000000001


My Empirical Probability for the picking of a fire type gave me 7.67% chance to get a fire type.

# Grass AND Poison Type Theoretical  Probability
## There are 9 pokemon which have both the types of grass and poison. Making the equation 9/152 = 0.059 or 5.9% chance to get a pokemon that is a grass and a poison type.

# Water OR Electric Type Empirical Probability

In [462]:
rounds = 0
EType = 0
WType = 0
trials = 10000
while rounds < 10000:
    num = random.randint(1,152)
    if num <= 32:
        WType += 1
    if num > 32 and num <=41:
        EType += 1
    rounds += 1

print(str(((EType+WType)/trials) * 100))

26.97


My Empirical Probability says there is a 26.97% chance to get a water OR electric type.

# My Answer
## My Empirical Probability for getting a fire type is a 7.67% chance.
## My Theoretical Probability for getting a grass or poison type is a 5.9% chance.
## My Empirical Probability for getting a water or electric type is a 26.97% chance.